### PR TITLE
[3.0] Changes WebApp to use Express under the hood

### PR DESCRIPTION
**Pull request by Grubba27**
_Sorry about the number of files and changes_

This PR focuses on changing connect framework for the Express framework.

## Why is this happening?

The way connect handles async handlers is kind of awkward because it was made at a different time than ours, while we were removing fibers we decided to do another upgrade in the MeteorJS core. Give a more modern-ish look for core functionality.

To make async code work properly you would need to do something like this in your codebase:
```js
 app.use(async function middleware1(req, res, next) {
    // middleware 1
   // if error occurs will rang
    const getFoo = await FooCol.find({}).fetchAsync();
    next();
  });

 //this is fine
 app.use(async function middleware2(req, res, next) {
    try{
      const getFoo = await FooCol.find({}).fetchAsync();
      next();
    } catch (e){next(e)}
  });

```

We could have a function like @radekmie has shown me:
```js
function wrap(fn) {
  return (request, response, next) => {
    Promise.resolve().then(fn).then(() => next(), next);
  };
}
```
and could be used like this to have the same functionality that we have in [Express](https://zellwk.com/blog/async-await-express/):

```js
WebApp.connectHandlers.use(wrap(async (request, response) => {
  // ...
}));
```

Now with the new API the code above can be replaced with the following:
```js
WebApp.expressHandlers.use(async (request, response) => {
  // ...
});
```

Does not seems much, but this will help us a lot in the core, and the community in the long term, as we can benefit a lot more from having 100% compatibility with the express plugin ecosystem.

## But express was built on top of connect why the change?
From express docs:

> Express 4 no longer depends on Connect, and removes all built-in middleware from its core, except for the express.static function. This means that Express is now an independent routing and middleware web framework

This is an overall improvement as we can do it all the same way as before but now our webapp package is a little bit more complete. 

  - `WebApp.connectHandlers.use(middleware)` is now `WebApp.expressHandlers.use(middleware)`
  - `WebApp.rawConnectHandlers.use(middleware)` is now `WebApp.rawExpressHandlers.use(middleware)`
  - `WebApp.connectApp` is now `WebApp.expressApp`

If you have an edge case that may break because of compatibility issues here is [express 4 migration guide](https://expressjs.com/en/guide/migrating-4.html)

---
#### Running against devel tests:
Screenshot 2023-01-16 at 10 25 07

---
This closes #12447

### CODE REVIEW

**Review by NicolasLopes7 — APPROVED** [review]

**Review by filipenevola — COMMENTED** [review]
It seems very interesting.

I would like to see this PR against current devel/master just to make sure everything is still working as expected.

Later we could apply it on top of 3.0 as I think this would be the right place to merge it.

What do you think @Grubba27 ?

**Comment by Grubba27** [comment]
Sounds great! 
I will try do this as soon as possible! 
I did tried running all tests, all but 1 ran smoothly, I did not waited it finish but I think it will be okay.

File: packages/webapp/webapp.d.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as http from 'http';
-import * as connect from 'connect';
+import e from '@types/express';
```
Comment (radekmie) [inline packages/webapp/webapp.d.ts:2]: You shuold never import `@types/*` directly:
```suggestion
import express from 'express';
```

**Review by radekmie — COMMENTED** [review]
My only concern about this change are the defaults of `connect` and `express`. For example, I think we definitely should opt out of `X-Powered-By`:
```ts
app.disable('x-powered-by');
```

And there are other things like that, including ETags. It may affect the performance and probably security _significantly_.

**Comment by Grubba27** [comment]
> And there are other things like that, including ETags. It may affect the performance and probably security significantly.

Can you enumerate a little bit more? I will make the adjustments with ```app.disable('x-powered-by');``` and ```app.set('etag', false);``` 
cc: @radekmie

**Comment by Grubba27** [comment]
Screenshot 2023-01-16 at 10 25 07

cc: @filipenevola

**Comment by zodern** [comment]
Replacing connect with something that has a router will be very nice.

It seems one of the main reasons for doing this now is express handles promise rejections better. This is coming in [express 5](https://expressjs.com/en/guide/migrating-5.html#rejected-promises), which is still in beta. As far as I know, express 4 doesn't handle it any better than connect does. We already [modify connect](https://github.com/meteor/meteor/blob/devel/packages/webapp/connect.js) so it supports fibers. We could do the same thing with connect or express so it handles rejected promises.

Since there are so many breaking changes in Meteor 3, I'm wondering if there is a way to switch to express in a way that doesn't add more. It is possible to use express as a middleware for connect. Instead of removing connect completely, we could keep it for now, and add `WebApp.expressHandlers` which runs after WebApp.rawConnectHandlers and maybe before WebApp.connectHandlers. Existing code could work as is until projects are ready to migrate, and new code can use express. Later connect could be deprecated and removed.

**Comment by Grubba27** [comment]
IMHO this:
 > We could do the same thing with connect or express, so it handles rejected promises.

It would be the best, but at the same time, I think that if we have the chance to move into something that is more well-known, it would be good for MeteorJS adoption.

Regarding the amount of changes that we will have on this 3.0, it seems overwhelming. At the same time, it would be hard to maintain both connect and express. Let's take this chance of having a breaking change and move a few pieces because if we cannot change now, when can we?

**Review by denihs — APPROVED** [review]

**Comment by radekmie** [comment]
> > And there are other things like that, including ETags. It may affect the performance and probably security significantly.
>
> Can you enumerate a little bit more? I will make the adjustments with `app.disable('x-powered-by');` and `app.set('etag', false);`

@Grubba27 First of all, I think it has to be done for every `express` instance, not only the top-level one. Secondly, I'm not sure there's anything else, but the best and easiest would be to run the app with and without this change, save `curl -v localhost:3000` result to a file and compare the headers and the rest of the output for a few endpoints (`/`, some dynamic import, a static file).

**Comment by Grubba27** [comment]
With express without disabling etag and x-powered-by:
Screenshot 2023-01-16 at 19 13 02


With connect:
Screenshot 2023-01-16 at 19 07 27

With express disabling etag and x-powered-by

Screenshot 2023-01-16 at 19 16 17

File: packages/webapp/webapp_server.js [hunk 2]
```diff
@@ -1039,12 +1036,19 @@ function runWebAppServer() {
   WebAppInternals.reloadClientPrograms();
 
   // webserver
-  var app = connect();
+  var app = express();
+
+  // Security and performace headers
+  // these headers come from these docs: https://expressjs.com/en/api.html#app.settings.table
+  app.set('x-powered-by', false);
+  app.set('etag', false);
```
[the remark this answers is not in the document]
Comment (Grubba27) [reply packages/webapp/webapp_server.js]: sounds good

**Comment by Grubba27** [comment]
Hey, @rj-david, I would say that it Is a good practice just take a look at express 4 migration just in case you use a very specific connect feature that was not ported to express, I will update the docs & PR description to reflect this

**Review by radekmie — APPROVED** [review]

**Review by henriquealbert — APPROVED** [review]
Great work as always @Grubba27 🎉

**Comment by StorytellerCZ** [comment]
Well #12410 happened much earlier than I expected. 😂

**Review by StorytellerCZ — APPROVED** [review]
